File: bffs/mobile-bff/app/clients/health_client.py
from datetime import datetime, timezone
import httpx

from app.config import USER_SERVICE_URL, RENTAL_SERVICE_URL, ITEM_HTTP_URL


async def check_http_service(name: str, url: str):
    try:
        async with httpx.AsyncClient(timeout=2) as client:
            response = await client.get(url)

        return {
            "name": name,
            "status": "UP" if 200 <= response.status_code < 300 else "DOWN",
            "details": response.json()
        }

    except Exception as ex:
        return {
            "name": name,
            "status": "DOWN",
            "error": str(ex)
        }


async def readiness():
    checks = [
        await check_http_service("user-service", f"{USER_SERVICE_URL}/health"),
        await check_http_service("item-service", f"{ITEM_HTTP_URL}/health"),
        await check_http_service("rental-service", f"{RENTAL_SERVICE_URL}/health"),
    ]

    status = "UP" if all(check["status"] == "UP" for check in checks) else "DEGRADED"

    return {
        "service": "mobile-bff",
        "status": status,
        "checks": checks,
        "timestamp": datetime.now(timezone.utc).isoformat()
    }

File: bffs/web-bff/src/clients/healthClient.js
const httpClient = require("./httpClient");
const env = require("../config/env");

async function checkHttpService(name, url) {
  try {
    const response = await httpClient.get(url, {
      timeout: 2000
    });

    return {
      name,
      status: response.status >= 200 && response.status < 300 ? "UP" : "DOWN",
      details: response.data
    };
  } catch (error) {
    return {
      name,
      status: "DOWN",
      error: error.message
    };
  }
}

async function readiness() {
  const checks = await Promise.all([
    checkHttpService("user-service", `${env.userServiceUrl}/health`),
    checkHttpService("item-service", `${env.itemHttpUrl}/health`),
    checkHttpService("rental-service", `${env.rentalServiceUrl}/health`)
  ]);

  const status = checks.every((check) => check.status === "UP") ? "UP" : "DEGRADED";

  return {
    service: "web-bff",
    status,
    checks,
    timestamp: new Date().toISOString()
  };
}

module.exports = {
  readiness
};